packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,436 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'image.dart';
+import 'ticker_provider.dart';
+
+/// An image that shows a [placeholder] image while the target [image] is
+/// loading, then fades in the new image when it loads.
+///
+/// Use this class to display long-loading images, such as [new Image.network],
+/// so that the image appears on screen with a graceful animation rather than a
+/// sudden jerk.
+///
+/// This image ignores the [placeholder]'s and [image]'s `color` and `blendMode`
+/// properties in order to efficiently perform the cross-fade animation.
+///
+/// If the [image]'s [ImageProvider] returns [ImageInfo] synchronously, such as
+/// when the image has been loaded and cached, the [image] is displayed
+/// immediately and the [placeholder] is never displayed.
+///
+/// [fadeOutDuration] and [fadeOutCurve] control the fade-out animation of the
+/// placeholder.
+///
+/// [fadeInDuration] and [fadeInCurve] control the fade-in animation of the
+/// target [image].
+///
+/// Prefer [placeholder] that's already cached so that it is displayed in one
+/// frame. This prevents it from appearing suddenly on the screen.
```

Comment:
Yeah, I can't come up with an explanation that's not a mouthful. The goal is for the placeholder to appear together with whatever widget it is embedded in in the same frame. It will appear immediately, but presumably the embedding widget already has a designer-approved transition applied to it and so the effect of having the placeholder appear suddenly won't appear abrupt to the user. As an example, if you are switching tabs, the tab content slides into view with an animation and if there's an image with a placeholder inside the content it will be animated into view as well. That's not the case for the long-loading image, which will appear after the transition is over.